internal/api/mfa.go

```diff
@@ -174,6 +257,83 @@ func (a *API) EnrollFactor(w http.ResponseWriter, r *http.Request) error {
 	})
 }
 
+func (a *API) challengeSMSFactor(w http.ResponseWriter, r *http.Request, user *models.User, factor *models.Factor) error {
+	ctx := r.Context()
+	config := a.config
+	db := a.db.WithContext(ctx)
+	ipAddress := utilities.GetIPAddress(r)
+	params := &ChallengeFactorParams{}
+	if err := retrieveRequestParams(r, params); err != nil {
+		return err
+	}
+	channel := params.Channel
+
+	if channel == "" {
+		channel = sms_provider.SMSProvider
+	}
+	if !sms_provider.IsValidMessageChannel(channel, config.Sms.Provider) {
+		return badRequestError(ErrorCodeValidationFailed, InvalidChannelError)
+	}
+
+	otp, err := crypto.GenerateOtp(config.MFA.SMS.OtpLength)
+	if err != nil {
+		panic(err)
+	}
+	challenge, err := factor.CreateSMSChallenge(ipAddress, otp, config.Security.DBEncryption.Encrypt, config.Security.DBEncryption.EncryptionKeyID, config.Security.DBEncryption.EncryptionKey)
+	if err != nil {
+		return internalServerError("error creating SMS Challenge")
+	}
+	message, err := generateSMSFromTemplate(config.MFA.SMS.SMSTemplate, otp)
+	if err != nil {
+		return internalServerError("error generating sms template").WithInternalError(err)
+	}
+	if config.Hook.SendSMS.Enabled {
+		input := hooks.SendSMSInput{
+			User: user,
+			SMS: hooks.SMS{
+				OTP:     otp,
+				SMSType: "mfa",
+			},
+		}
+		output := hooks.SendSMSOutput{}
+		err := a.invokeHook(a.db, r, &input, &output, a.config.Hook.SendSMS.URI)
+		if err != nil {
+			return internalServerError("error invoking hook")
+		}
+	} else {
+		if !challenge.SentAt.Add(config.MFA.SMS.MaxFrequency).Before(time.Now()) {
+			return MaxFrequencyLimitError
+		}
+		smsProvider, terr := sms_provider.GetSmsProvider(*config)
+		if terr != nil {
+			return internalServerError("Unable to get SMS provider").WithInternalError(terr)
+		}
```

Comment:
i think it's better to check for this at the start rather than creating the challenge in the db which would be orphaned if there is no sms provider set up??

Reply:
yup, good catch